cla_backend/apps/cla_eventlog/tests/test_find_and_delete_old_cases.py

```diff
@@ -0,0 +1,152 @@
+from django.test import TestCase
+from django.utils import timezone
+from freezegun import freeze_time
+import datetime
+
+from core.tests.mommy_utils import make_recipe
+from cla_eventlog.models import Log
+from complaints.models import Complaint
+from legalaid.models import Case, EODDetails, EligibilityCheck, PersonalDetails
+from diagnosis.models import DiagnosisTraversal
+from cla_auditlog.models import AuditLog
+from cla_eventlog.management.commands.find_and_delete_old_cases import Command
+
+
+def _make_datetime(year=None, month=None, day=None, hour=0, minute=0, second=0):
+    today = datetime.date.today()
+    year = year if year else today.year
+    month = month if month else today.month
+    day = day if day else today.day
+    dt = datetime.datetime(year, month, day, hour, minute, second)
+    return timezone.make_aware(dt, timezone.get_current_timezone())
+
+
+class FindAndDeleteOldCases(TestCase):
+    def create_case(self, current_time, case_type="legalaid.case"):
+        case = None
+        freezer = freeze_time(current_time)
+        freezer.start()
+        case = make_recipe(case_type)
+        freezer.stop()
+        return case
+
+    def create_event_log_for_case(self, case, code, created):
+        make_recipe("cla_eventlog.log", case=case, code=code, created=created)
+
+    def delete_old_cases(self, current_time):
+        freezer = freeze_time(current_time)
+        freezer.start()
+        Command().execute()
+        freezer.stop()
+
+    def test_old_case_with_recent_event_logs(self):
+        date = _make_datetime(year=2014, month=4, day=27, hour=9)
+        case = self.create_case(date)
+
+        self.create_event_log_for_case(case, "CASE_VIEWED", _make_datetime(year=2014, month=5, day=30, hour=9))
+        self.create_event_log_for_case(case, "CASE_VIEWED", _make_datetime(year=2018, month=4, day=27, hour=9))
+        self.create_event_log_for_case(case, "CASE_VIEWED", _make_datetime(year=2020, month=4, day=27, hour=9))
+
+        self.delete_old_cases(_make_datetime(year=2021, month=1, day=1, hour=9))
+
+        self.assertEqual(Case.objects.all().count(), 1)
```

Comment:
You don't have to use `.all()` if you're going to be chaining something like `.count()` on

It's enough for you to know at this point that that's generally the rule, and that it applies to everything built-in, like `.count()`, `.filter()`, `.exclude()`, `.first()`, `.update()` etc etc

The technical reason is that all of those are defined on the queryset class (which is what `.all()` returns) but are also deliberately made available on the manager class too (which is what `objects` is), specifically so that they can be chained but don't _have_ to be.

So in all cases where you have `objects.all().count()` you can simplify this to `objects.count()`; it won't actually have any effect on the SQL that is run, because it will already be being processed to the exact same thing.

Reply:
Roger that

Reply:
As I say above, strictly speaking no. But all of the ones provided by Django (as opposed to added by us or some third-party library), yes